### ISSUE TITLE
Windows High Contrast Mode / forced color adaptations

### ISSUE THREAD

**Issue by patrickhlauke** [twbs/bootstrap#35941]
### Prerequisites

- [X] I have [searched](https://github.com/twbs/bootstrap/issues?utf8=%E2%9C%93&q=is%3Aissue) for duplicate or closed feature requests
- [X] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)

### Proposal

Adding a few surgical, very specific general styles to Bootstrap using the `(forced-colors)` media query, to make sure that some of our components that aren't properly discernible in Windows High Contrast Mode stand a slightly better chance at having at least minimal styles applied even in WHCM.

### Motivation and context

Splitting the idea out of https://github.com/twbs/bootstrap/pull/29490

A lot of the standard Bootstrap styles rely on background colours, borders and shadows (e.g. to make buttons/pills/etc look "button-like"). These styles usually get completely overridden by Windows High Contrast Mode (WHCM), which explicitly ignores things like background and forces specific user-defined colours for things.

Until recently, there was no clean standardised way of approaching this issue (even with proprietary things like `-ms-high-contrast` - see https://www.tpgi.com/windows-high-contrast-mode-the-limited-utility-of-ms-high-contrast/)

However, the relatively new `(forced-colors)` media query offers a relatively stable way of approaching this issue now, adding very targeted adaptations/tweaks for cases where WHCM is enabled.

I have recently been playing with it on a project (not directly using Bootstrap, but the concept is obviously the same) to force an outline (rather than a border, so it doesn't affect element dimensions), a la

```
@media (forced-colors) {
  button {
    outline: 1px solid LinkText !important;
  }
}
```

Trying to do further/more intense styling is probably counterproductive (i.e. not doing a whole exercise like dark mode or anything), but just some basics like this would immediately make BS more WHCM-friendly for very little extra effort.

**Comment by davidscotson** [comment]
Hi  @patrickhlauke, are you aware of a shared gist or other source that gathers together the forced-color adaptions that aren't yet accepted into upstream Bootstrap? I'm looking at adding them to a Bootstrap 4(.6) project and if one doesn't exist I might create it as it seems generally useful and shareable, rather than project specific.

**Comment by patrickhlauke** [comment]
i don't think there's a single collated list at the moment. would suggest, if you can find the time, going through the docs pages and trying to identify all the different places where currently forced colors/WHCM is not ideal (in terms of borders, outlines, etc not showing up)

**Comment by davidscotson** [comment]
I've started a gist here, I'll try to remember to update it as I test and find issues:

https://gist.github.com/davidscotson/446f5a637715af9d1db832f60640e705

**Comment by patrickhlauke** [comment]
Taking https://getbootstrap.com/docs/5.2/examples/cheatsheet/ for a quick spin in high contrast (Windows 10 WHCM, Chrome):

# Forms

## Overview

* checked checkbox, selected radio button, shows no focus indication
![image](https://user-images.githubusercontent.com/895831/181859416-e7712a2c-a438-4ff6-9610-9adb27113618.png)


* switch checkbox input: when not checked, only shows outline border; when focused, shows faint off switch; when checked, has no focus indicator
![image](https://user-images.githubusercontent.com/895831/181858852-69962704-a566-439d-932a-529e3a8a0f9c.png)
![image](https://user-images.githubusercontent.com/895831/181858889-a017deeb-de37-4947-8e76-d581342a9e3f.png)
![image](https://user-images.githubusercontent.com/895831/181858951-a471ebde-b453-4405-a4c9-86d218292fcf.png)

* example range totally invisible
![image](https://user-images.githubusercontent.com/895831/181859206-6d189d42-ed4d-4620-92a3-e0e978079389.png)

## Floating labels

* floating label doesn't hide placeholder text; when focused, shows awkward focus outline
![image](https://user-images.githubusercontent.com/895831/181859196-71dcbe14-35bc-48a0-b462-91f65411124a.png)

## Validation

* "Agree..." checkbox shows no focus indication
![image](https://user-images.githubusercontent.com/895831/181859179-9854d368-1654-4b28-96a6-40f3937935c9.png)


# Components

* Accordion shows no focus indication

![image](https://user-images.githubusercontent.com/895831/181859163-4a05cfe9-89ad-4545-9751-ad178f234587.png)


## Alerts

* No visible X close control, no focus indication for it
![image](https://user-images.githubusercontent.com/895831/181859101-b9b1dc7a-6d2d-4356-a8cd-6cfd58e2ecd0.png)


## Badge

* Shows no outline/border
![image](https://user-images.githubusercontent.com/895831/181859002-de05b02b-82b5-441d-9541-2845e109af8a.png)


## Carousel

* Dots look awkward/wrong
![image](https://user-images.githubusercontent.com/895831/181859456-546cb336-81c1-4b53-afbc-6dd12b566b84.png)

## Dropdowns

* Down arrow looks wrong
* Focus indication on active dropdown item not shown
![image](https://user-images.githubusercontent.com/895831/181859318-5cf15ff0-5316-48ca-b1e9-86a8a6940e2d.png)


## Modal

* No visible X close control, no focus indication for it
![image](https://user-images.githubusercontent.com/895831/181859498-855df3ea-dd7c-4cbb-be76-14a89f472413.png)

## Navs

* doesn't show active one
![image](https://user-images.githubusercontent.com/895831/181858725-202e569a-59e8-4318-983d-49151f16aeaf.png)

## Navbar

* First example, logo doesn't show
![image](https://user-images.githubusercontent.com/895831/181859275-8f237361-741a-40bf-b6e8-c1b4b789dc4d.png)

## Pagination

* Current active one not shown
![image](https://user-images.githubusercontent.com/895831/181859253-87418f59-80e2-4282-8a15-4ba9b943f664.png)


## Popovers

* Arrow looks awkward/wrong
![image](https://user-images.githubusercontent.com/895831/181859262-6f0c9410-52d2-4c90-b9f1-c69d6aaf6d78.png)

## Progress

* Bar is totally invisible
![image](https://user-images.githubusercontent.com/895831/181859242-b956d75d-771a-4566-a02d-490e10cfbe1a.png)

## Spinners

* Completely static circles
![image](https://user-images.githubusercontent.com/895831/181859239-612b7f98-93e8-4768-b309-7949a5814a2d.png)


## Toasts

* No visible X close control, no focus indication for it
![image](https://user-images.githubusercontent.com/895831/181859232-3b9c5e9c-c9f3-4f34-ba50-3b69b8302d6c.png)


## Tooltips

* Arrow looks awkward/wrong
![image](https://user-images.githubusercontent.com/895831/181859223-212bf331-3f60-4c8e-aa23-c0559b731199.png)


A few other examples not covered in the cheatsheet:

* toggle buttons https://getbootstrap.com/docs/5.2/components/buttons/#toggle-states don't visually differentiate active one
![image](https://user-images.githubusercontent.com/895831/181859558-46279dea-476b-467a-9dea-f6379f4fe5bc.png)

* checkbox/radio button group https://getbootstrap.com/docs/5.2/components/button-group/#checkbox-and-radio-button-groups don't differentiate checked/unchecked
![image](https://user-images.githubusercontent.com/895831/181859750-e19e38ef-780e-403d-96fb-d5ae69752c1a.png)

*  as already seen in the alert and modal above, close button https://getbootstrap.com/docs/5.2/components/close-button/ doesn't show at all (except for its white variant)
![image](https://user-images.githubusercontent.com/895831/181859846-5d33f5dd-0d0a-42b5-ab7c-b9a456fbac2f.png)

**Comment by robfentress** [comment]
Another thing I didn't see in previous comments is the fact that you can't see the `checked` state of the checkboxes when viewed in certain themes, such as High Contrast White in Windows 10 and  Desert High Contrast in Windows 11.  You can find an explanation of the issues I found with this control and my proposed solutions in this CodePen:
[Fixing Accessibility of Bootstrap's Checkbox Styling](https://codepen.io/rfentress/pen/jOpEbYa)

This is easy to miss, because it doesn't show up when checking with darker themes, which is all most people test with.

I'm not sure, but I think I remember there being similar issues with the radio buttons.

**Comment by coliff** [comment]
I spent a lot of time going through the components and have created a `forced-colors: active` stylesheet which fixes just about all issues I could find in both dark and light high contrast themes. Tested with Edge on Windows 11.

- GitHub repo: https://github.com/coliff/bootstrap-forced-colors-css
- npm: https://www.npmjs.com/package/bootstrap-forced-colors-css
- Demo: https://coliff.github.io/bootstrap-forced-colors-css/tests/

Some fixes made are a bit opinionated, e.g. increasing border width from 1px to 2px on form inputs to counter the fact that there is no box-shadow. It'd be great to see some/most of these fixes be made to Bootstrap in the future. For now though, users can use this stylesheet to improve things for Windows users with High Contrast themes.

```html
<link
  rel="stylesheet"
  href="https://cdn.jsdelivr.net/npm/bootstrap-forced-colors-css@1/css/bootstrap-forced-colors.min.css" 
  media="screen and (forced-colors: active)">
```

<details>
<summary>Testing with Forced Colors Mode Emulation in Chrome / Edge</summary>

![edge-test-forced-colors-active](https://github.com/twbs/bootstrap/assets/1212885/dc681871-7605-42d8-82d0-b6361e7f3faa)

![edge-test-forced-colors-active-light-mode](https://github.com/twbs/bootstrap/assets/1212885/df7766a8-d825-4845-8a2e-1f733dbd38e8)

</details>
